packages/react/src/reactrouterv6.tsx

```diff
@@ -0,0 +1,162 @@
+import { Transaction, TransactionContext } from '@sentry/types';
+import hoistNonReactStatics from 'hoist-non-react-statics';
+import React from 'react';
+
+import { Action, Location } from './types';
+
+interface RouteObject {
+  caseSensitive?: boolean;
+  children?: RouteObject[];
+  element?: React.ReactNode;
+  index?: boolean;
+  path?: string;
+}
+
+type Params<Key extends string = string> = {
+  readonly [key in Key]: string | undefined;
+};
+
+interface RouteMatch<ParamKey extends string = string> {
+  params: Params<ParamKey>;
+  pathname: string;
+  route: RouteObject;
+}
+
+type UseEffect = (cb: () => void, deps: unknown[]) => void;
+type UseLocation = () => Location;
+type UseNavigationType = () => Action;
+type CreateRoutesFromChildren = (children: JSX.Element[]) => RouteObject[];
+type MatchRoutes = (routes: RouteObject[], location: Location) => RouteMatch[] | null;
+
+let _useEffect: UseEffect;
+let _useLocation: UseLocation;
+let _useNavigationType: UseNavigationType;
+let _createRoutesFromChildren: CreateRoutesFromChildren;
+let _matchRoutes: MatchRoutes;
+let _customStartTransaction: (context: TransactionContext) => Transaction | undefined;
+let _startTransactionOnLocationChange: boolean;
+let _startTransactionOnPageLoad: boolean;
+
+export function reactRouterV6Instrumentation(
+  useEffect: UseEffect,
+  useLocation: UseLocation,
+  useNavigationType: UseNavigationType,
+  createRoutesFromChildren: CreateRoutesFromChildren,
+  matchRoutes: MatchRoutes,
+) {
+  return (
+    customStartTransaction: (context: TransactionContext) => Transaction | undefined,
+    startTransactionOnPageLoad = true,
+    startTransactionOnLocationChange = true,
+  ): void => {
+    _useEffect = useEffect;
+    _useLocation = useLocation;
+    _useNavigationType = useNavigationType;
+    _matchRoutes = matchRoutes;
+    _createRoutesFromChildren = createRoutesFromChildren;
+
+    _customStartTransaction = customStartTransaction;
+    _startTransactionOnLocationChange = startTransactionOnLocationChange;
+    _startTransactionOnPageLoad = startTransactionOnPageLoad;
+  };
+}
+
+const getTransactionName = (routes: RouteObject[], location: Location, matchRoutes: MatchRoutes): string => {
+  if (!routes || routes.length === 0 || !matchRoutes) {
+    return location.pathname;
+  }
+
+  const branches = matchRoutes(routes, location);
+
+  if (branches) {
+    // eslint-disable-next-line @typescript-eslint/prefer-for-of
+    for (let x = 0; x < branches.length; x++) {
+      if (branches[x].route && branches[x].route.path && branches[x].pathname === location.pathname) {
+        return branches[x].route.path || location.pathname;
+      }
+    }
+  }
+
+  return location.pathname;
+};
+
+const SENTRY_TAGS = {
+  'routing.instrumentation': 'react-router-v6',
+};
+
+export function withSentryV6<P extends Record<string, any>, R extends React.FC<P>>(Routes: R): R {
```

Comment:
Let's call this `withSentryReactRouterV6Routing`. It's way longer, but it makes it very clear what is happening to users (+ will all get minified in the end anyway).